subsys/bluetooth/mesh/net.c

```diff
@@ -648,101 +650,197 @@ static bool net_decrypt(struct bt_mesh_net_rx *rx, struct net_buf_simple *in,
 		return;
 	}
 
-	if (rx->net_if == BT_MESH_NET_IF_ADV &&
-	    !rx->friend_cred &&
-	    bt_mesh_relay_get() != BT_MESH_RELAY_ENABLED &&
-	    bt_mesh_gatt_proxy_get() != BT_MESH_GATT_PROXY_ENABLED) {
+	buf = bt_mesh_adv_main_create(BT_MESH_ADV_DATA,
+				      bt_mesh_net_transmit_get(), K_NO_WAIT);
+	if (!buf) {
+		LOG_DBG("Out of relay buffers");
 		return;
 	}
 
-	LOG_DBG("TTL %u CTL %u dst 0x%04x", rx->ctx.recv_ttl, rx->ctl, rx->ctx.recv_dst);
+	(void)net_buf_simple_add_mem(&buf->b, sbuf->data, sbuf->len);
+
+	cred = &rx->sub->keys[SUBNET_KEY_TX_IDX(rx->sub)].msg;
 
-	/* The Relay Retransmit state is only applied to adv-adv relaying.
-	 * Anything else (like GATT to adv, or locally originated packets)
-	 * use the Network Transmit state.
+	/* Update NID if RX or RX was with friend credentials */
+	if (rx->friend_cred) {
+		buf->b.data[0] &= 0x80; /* Clear everything except IVI */
+		buf->b.data[0] |= cred->nid;
+	}
+
+	/* When the Friend node relays message for lpn, the message will be
+	 * retransmitted using the managed flooding security credentials and
+	 * the Network PDU shall be retransmitted to all network interfaces.
+	 *
+	 * If the Proxy feature is supported and enabled, the Network PDU shall
+	 * be retransmitted to all network interfaces.
 	 */
-	if (rx->net_if == BT_MESH_NET_IF_ADV && !rx->friend_cred) {
-		transmit = bt_mesh_relay_retransmit_get();
-	} else {
-		transmit = bt_mesh_net_transmit_get();
+	bt_mesh_adv_send(buf, NULL, NULL);
+
+	if (IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY)) {
+		bt_mesh_proxy_relay(buf, rx->ctx.recv_dst);
+	}
+
+	bt_mesh_buf_unref(buf);
+}
+#endif /* CONFIG_BT_MESH_FRIEND || CONFIG_BT_MESH_PROXY */
+
+#if defined(CONFIG_BT_MESH_RELAY)
+static void outbound_bearer_adv(struct net_buf_simple *sbuf,
+				struct bt_mesh_net_rx *rx)
+{
+	struct bt_mesh_buf *buf;
+	uint8_t prio = 0;
+
+	/* If the Relay feature is supported and enabled, the Network PDU shall be tagged
+	 * as relay, and the Network PDU shall be retransmitted to all network interfaces
+	 * connected to the advertising bearer.
+	 */
+	if (bt_mesh_relay_get() != BT_MESH_RELAY_ENABLED) {
+		return;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_MESH_RELAY_PRIORITY) && prio_cb) {
+		prio = prio_cb(&rx->ctx);
 	}
 
-	buf = bt_mesh_adv_create(BT_MESH_ADV_DATA, BT_MESH_RELAY_ADV,
-				 transmit, K_NO_WAIT);
+	buf = bt_mesh_adv_relay_create(prio, bt_mesh_relay_retransmit_get());
 	if (!buf) {
 		LOG_DBG("Out of relay buffers");
 		return;
 	}
 
+	(void)net_buf_simple_add_mem(&buf->b, sbuf->data, sbuf->len);
+
+	bt_mesh_adv_send(buf, NULL, NULL);
+
+	bt_mesh_buf_unref(buf);
+}
+#endif /* CONFIG_BT_MESH_RELAY */
+
+#if defined(CONFIG_BT_MESH_PROXY)
+static void outbound_bearer_proxy(struct net_buf_simple *sbuf,
+				  struct bt_mesh_net_rx *rx)
+{
+	struct bt_mesh_buf *buf;
+
+	if (bt_mesh_gatt_proxy_get() != BT_MESH_GATT_PROXY_ENABLED) {
+		return;
+	}
+
+	buf = bt_mesh_adv_main_create(BT_MESH_ADV_DATA,
```

Comment:
I'm not sure I like this. Previously we didn't use an extra buffer. Now we don't only use an extra buffer, but also the local pool for relay messages. This will also occupy the whole local pool in high density traffic environment. And I think my comment https://github.com/zephyrproject-rtos/zephyr/pull/52072#discussion_r1052129083 is still relevant. This creates a situation where messages relayed to proxy, but not to adv bearer.

Maybe we could tag proxy packets somehow and then skip them?

Reply:
Well, I think most high density situations is adv to adv, not proxy, also for relay definitions, We confused these concepts before, I solved it by the way this time

Reply:
Also according https://github.com/zephyrproject-rtos/zephyr/pull/52072#discussion_r1044054258

The message delivered from adv bearer, if proxy support and enable, should relay to gatt bearer, this 'relay' not 'BT_MESH_RELAY', also shouldn't occupy relay buffer.

Reply:
BTW: the bt_mesh_relay function In the previous design, this function has always existed, even if there is no relay support.

Reply:
By the way, I remember that in the implementation of zephyr, the Bluetooth GATT connection also has a buffer, that is, it does not support intensive packet sending.

Reply:
> Well, I think most high density situations is adv to adv, not proxy(because it's need keep a gatt connection), also for relay definitions, We confused these concepts before, I solved it by the way this time

If you have node A with gatt proxy connection and node B that sends or relays enormous amount of traffic, node A will relay those messages (assuming they bypassed all checks and filters). We can't disregard this case if prioritization feature is enabled on a node that supports GATT Proxy.

> The message delivered from adv bearer, if proxy support and enable, should relay to gatt bearer, this 'relay' not 'BT_MESH_RELAY', also shouldn't occupy relay buffer.

My main point is previously we were more efficient in terms of memory usage as we were using the same buffer and were just incrementing the reference. Now, as I mentioned above, we will use both pools. The spec doesn't care about buffers.

> the bt_mesh_relay function In the previous design, this function has always existed, even if there is no relay support.

I didn't get this point.

> By the way, I remember that in the implementation of zephyr, the Bluetooth GATT connection also has a buffer, that is, it does not support intensive packet sending.

I guess it doesn't matter as long as a user can configure the mesh with prioritization feature and GATT Proxy.

----

One more thing that came up to me regarding the previous code. In case when adv_ext.c relayed message before proxy, dereferencing the buffer will brake the code once the message is relayed by proxy. Since we can't cancel bt_gatt_notify_cb call, we can't touch that message, otherwise in proxy we will dereference either a new buf or buf with ref == 0.

Tbh at the moment I don't see a good way around this except either tagging messages relayed over gatt and skipping them (thought I expect it won't work if all messages are relayed over gatt), or making the prioritization and gatt proxy features mutually exclusive.

Reply:
> If you have node A with gatt proxy connection and node B that sends or relays enormous amount of traffic, node A will relay those messages (assuming they bypassed all checks and filters). We can't disregard this case if prioritization feature is enabled on a node that supports GATT Proxy.

Also have another solution, we could add new inner functions that send proxy relay messages without buf.

Reply:
> I didn't get this point.

I'm mean in the current implementation (main branch), even we do not support relay and proxy and friend, the bt_mesh_relay function also be compiled in final binary firmware.

Reply:
> Also have another solution, we could add new inner functions that send proxy relay messages without buf.

Right, buf itself is not really needed for proxy, but we still need to notify upper layer about finishing transmission.

Reply:
One another way is send adv firstly, and send proxy relay message in adv:send_start.

Reply:
But will try add more args.

Reply:
Maybe can reuse buf.b.size or buf.b.__bufs

Reply:
@PavelVPV Ping

Reply:
> One another way is send adv firstly, and send proxy relay message in adv:send_start.

I didn't get how this will help. We need to keep adv buf in pool until we finished transmission over GATT. If the adv buf that we want to remove should be relayed over gatt, and we scheduled its transmission, we can't remove it from the pool.

> Maybe can reuse buf.b.size or buf.b.__bufs

Could you elaborate this?

Reply:
I still stick to the existing solution(base on relay mapping table), because the Mesh Profile defines a set of relay concepts, including proxy and friend, and these concepts were confused in the previous code implementation, resulting in the current problem. The existing separation design is for the future The maintenance is better, even if it takes up one more buffer.